core-bundle/src/Csrf/CsrfTokenManager.php

```diff
@@ -0,0 +1,120 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Csrf;
+
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\Security\Csrf\CsrfToken;
+use Symfony\Component\Security\Csrf\CsrfTokenManagerInterface;
+use Symfony\Component\Security\Core\Exception\RuntimeException;
+use Symfony\Component\HttpFoundation\Session\SessionInterface;
+use Symfony\Component\Security\Csrf\Exception\TokenNotFoundException;
+
+class CsrfTokenManager implements CsrfTokenManagerInterface
+{
+    /**
+     * @var SessionInterface
+     */
+    private $session;
+
+    /**
+     * @var RequestStack
+     */
+    private $requestStack;
+
+    /**
+     * @var string
+     */
+    private $secret;
+
+    public function __construct(SessionInterface $session, RequestStack $requestStack, string $secret)
+    {
+        $this->session = $session;
+        $this->requestStack = $requestStack;
+        $this->secret = $secret;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getToken($tokenId)
+    {
+        return new CsrfToken(
+            $tokenId,
+            $this->generateHash($tokenId, $this->getUserId())
+        );
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function isTokenValid(CsrfToken $token)
+    {
+        return
```

Comment:
This would be a security issue right? I can just attack by using the IP-bound token. It would validate even if the user is currently authenticated using a session.

Reply:
Both hashes use the session ID if present, one takes the value from the request the other from the currently started session, which may differ because the session ID could have been regenerated.

Should be no security issue, but thinking again about it we probably have to only compare the session ID from the request because that is the only one that could have been used to generate the token...

Reply:
Done in 94287a9ddf1d61529c78327b35807354c20190aa